consensus/src/main/java/org/apache/iotdb/consensus/multileader/logdispatcher/LogDispatcher.java

```diff
@@ -0,0 +1,284 @@
+        batch = new PendingBatch(startIndex, endIndex, logBatches);
+        logger.debug("accumulated a {} from wal", batch);
+      } else {
+        Iterator<IndexedConsensusRequest> iterator = bufferedRequest.iterator();
+        IndexedConsensusRequest prev = iterator.next();
+        // Prevents gap between logs. For example, some requests are not written into the queue when
+        // the queue is full. In this case, requests need to be loaded from the WAL
+        constructBatchFromWAL(startIndex, prev.getSearchIndex(), logBatches);
+        constructBatchIndexedFromConsensusRequest(prev, logBatches);
+        endIndex = prev.getSearchIndex();
+        iterator.remove();
+        while (iterator.hasNext()
+            && logBatches.size() <= MultiLeaderConsensusConfig.MAX_REQUEST_PER_BATCH) {
+          IndexedConsensusRequest current = iterator.next();
+          // Prevents gap between logs. For example, some logs are not written into the queue when
+          // the queue is full. In this case, requests need to be loaded from the WAL
+          if (current.getSearchIndex() != prev.getSearchIndex() + 1) {
+            constructBatchFromWAL(prev.getSearchIndex(), current.getSearchIndex(), logBatches);
+          }
+          constructBatchIndexedFromConsensusRequest(current, logBatches);
+          endIndex = current.getSearchIndex();
+          prev = current;
+          // We might not be able to remove all the elements in the bufferedRequest in the
+          // current function, but that's fine, we'll continue processing these elements in the
+          // bufferedRequest the next time we go into the function, they're never lost
+          iterator.remove();
+        }
+        batch = new PendingBatch(startIndex, endIndex, logBatches);
+        logger.debug("accumulated a {} from queue and wal", batch);
+      }
+      return batch;
+    }
+
+    public void sendBatchAsync(PendingBatch batch, DispatchLogHandler handler) {
+      if (!stopped) {
+        try {
+          AsyncMultiLeaderServiceClient client = clientManager.borrowClient(peer.getEndpoint());
+          TSyncLogReq req =
+              new TSyncLogReq(peer.getGroupId().convertToTConsensusGroupId(), batch.getBatches());
+          client.syncLog(req, handler);
+        } catch (IOException | TException e) {
+          logger.error("Can not sync logs to peer {} because", peer, e);
+        }
+      }
+    }
+
+    public SyncStatus getSyncStatus() {
+      return syncStatus;
+    }
+
+    private long constructBatchFromWAL(
+        long currentIndex, long maxIndex, List<TLogBatch> logBatches) {
+      while (currentIndex < maxIndex
+          && logBatches.size() <= MultiLeaderConsensusConfig.MAX_REQUEST_PER_BATCH) {
+        // TODO iterator
+        IConsensusRequest data = reader.getReq(currentIndex++);
+        if (data != null) {
+          ByteBuffer buffer = ByteBuffer.allocate(DEFAULT_BUFFER_SIZE);
```

Comment:
What if this buffer can't hold this request? Maybe we should use some sort of stream to handle this.

Reply:
We plan to refactor the serialization logic for the new architecture, which is expected to start next week. I forgot to add todo here